proof/final_exp.saw

```diff
@@ -0,0 +1,306 @@
+      } else do {
+        assume_unsat;
+      };
+    });
+
+// x * y == y * x
+mul_commutes_fp12_thm <- prove_cryptol {{ \x y  -> Fp_12.mul (x, y) == Fp_12.mul (y,x) }} [];
+
+// If x is in the cyclotomic subgroup, then fp12_conjugate (x^e) == x^-e.
+// Proof sketch:
+//   If x is in the cyclotomic subgroup,  then
+//     x^(p^4-p^2+1) = 1
+//     so x^(p^6+1) = x^( (p^4-p^2+1) * (p^2 +1) ) = 1^(p^2+1) = 1
+//     But then x^(p^6) * x = x^(p^6+1) = 1, so x^(p^6) is x inverse.
+fp12_conjugate_inverse_thm <- admit_cryptol
+  {{ \b e -> fp12_conjugate (C_expt b e) ==
+             if is_cyclotomic b
+             then C_expt b (-1 * e)
+             else apply fp12_conjugate (C_expt b e) }};
+
+// Translate k_expt to C_expt.  This can always be done because k_expt only
+// handles positive exponents, and C_expt supports all `b` values with positive
+// exponents.
+k_expt_C_expt_thm <- admit_cryptol
+  {{ \b e -> k_expt`{5} Fp_12 b e == C_expt b (toInteger e) }};
+
+// The first part of the final exponentiation function raises x to
+// (p^6-1)*(p^2+1).  This produces a value in the cyclotomic subgroup.
+final_exp_frobenius_map_is_cyclotomic_thm <- admit_cryptol
+  {{ \x -> is_cyclotomic (final_exp_frobenius_map x) == True }};
+
+// If b is in the cyclotomic subgroup and e1 == e2 (mod p^4-p^2+1), then
+// b^e1 == b^e2.
+log_cyclotomic_thm <- admit_cryptol
+  {{ \b e1 e2 -> Fp_12.is_equal (C_expt b e1, C_expt b e2) ==
+                 if is_cyclotomic b
+                 then (e1 % `cyclotomic_exponent_modulus) ==
+                      (e2 % `cyclotomic_exponent_modulus)
+                 else apply Fp_12.is_equal (C_expt b e1, C_expt b e2) }};
+
+// x * (1 / y) == x / y
+mul_inverse_div_fp12_ov <- admit_cryptol
+  {{ \x y -> Fp_12.mul (x, Fp_12.div (Fp_12.field_unit, y)) == Fp_12.div (x, y) }};
+
+// If x is in the cyclotomic subgroup, then x^(p^4-p^2+1) == 1.
+cyclotomic_unit_thm <- prove_cryptol (rewrite fp12_simpset
+  {{ \x -> C_expt x `cyclotomic_exponent_modulus ==
+           if is_cyclotomic x
+           then Fp_12.field_unit
+           else apply C_expt x `cyclotomic_exponent_modulus }})
+  fp12_unints;
+
+// forall x and e: x^(((p^6-1)*(p^2+1)))^e is in the cyclotomic subgroup.
+expt_final_exp_frobenius_map_is_cyclotomic_thm <- custom_prove_cryptol
+  {{ \x e -> is_cyclotomic (C_expt (final_exp_frobenius_map x) e) == True }}
+  do {
+    unfolding ["is_cyclotomic"];
+    simplify fp12_simpset;
+    simplify (addsimps [ cyclotomic_unit_thm
+                       , is_cyclotomic_expt_thm
+                       , final_exp_frobenius_map_is_cyclotomic_thm ] empty_ss);
+    w4_unint_z3 fp12_unints;
+  };
+
+// (if True then x else y) == x
+if_true_fp12 <- parse_core "\\ (x:(Vec 2 (Vec 3 (Vec 2 Integer)))) -> \\ (y:(Vec 2 (Vec 3 (Vec 2 Integer)))) -> ite (Vec 2 (Vec 3 (Vec 2 Integer))) True x y";
+elim_if_true_thm <- prove_cryptol (rewrite (cryptol_ss()) {{ \x y -> if_true_fp12 x y == x }}) [];
```

Comment:
FYI, `basic_ss` has a rule that does this and a few other propositional simplifications.